Print extra information about compiler C++11 compatibility.

jwpeterson opened the pull request:
This implements most of what is needed for #6564.

For instance, on Clang 3.7.0 at the end of the simulation, it now prints

```
Compiler C++11 compatibility report: 
  Basic functionality:     yes
  Deleted functions:       yes
  'final' keyword:         yes
  Initializer lists:       yes
  'move' keyword:          yes
  'nullptr' keyword:       yes
  'override' keyword:      yes
  <regex>:                 yes
  std::shared_ptr:         yes
  <thread>:                yes
  <type_traits>:           yes
  std::unique_ptr:         yes
  Variadic templates:      yes
```

while on GCC 4.6.3, a compiler that does not support much C++11 stuff, it prints:

```
Compiler C++11 compatibility report: 
  Basic functionality:     no
  Deleted functions:       yes
  'final' keyword:         no
  Initializer lists:       no
  'move' keyword:          no
  'nullptr' keyword:       no
  'override' keyword:      no
  <regex>:                 no
  std::shared_ptr:         no
  <thread>:                no
  <type_traits>:           no
  std::unique_ptr:         no
  Variadic templates:      yes
```

What is still needed is:
- [x] A command line argument that will turn off this printing altogether.  I need suggestions for what this should be...
- [x] If the user explicitly disables C++11 support by configuring with --disable-cxx11, all of the features above will currently print "no" because they won't be tested.  I can fix this so that it runs the tests even if you explicitly disable C++11 support, and therefore still prints a useful feature summary, but that will require a libmesh update (this is implemented in libMesh/libmesh#869).

This PR also contains a libmesh update.  The following is a short summary of changes to libmesh:
- Enable FETypes of arbitrary order (larger than the largest Order enumeration).
- Properly free communicators created by calling split().
- Fixes for adaptive p and h-p refinement.
- More thorough testing for C++11 threads (std::atomic, std::mutex, etc).
- Fix issue with VTK autoconf tests that caused test programs to fail linking.
- Enhance and expand C++11 configure tests.
- Make TypeTensor row() const.
- Optimize/fix issues with constant monomial projection.
- Make --enable-unique-ptr is now the default, MOOSE has been using this for a while already.
- Allow threading directly over std::vectors (without copying).
- Fix issue with XDA I/O on highly-refined meshes.

jwpeterson committed:
Print extra information about compiler C++11 compatibility.

* Add --check-cxx11 command line option.  The extra information is
  only printed if you pass this flag to the executable.  Note that you
  can simply run ./moose_test-opt --check-cxx11 without an input file
  and get the printout.
* Also write C++11 compatibility information to file, and print
  instructions for sending it to moose-users.
* This reports on several new C++11 feature checks that were added
  recently, therefore it requires a libmesh update to work correctly.

Refs #6564.

permcody commented on framework/src/base/MooseApp.C:
```diff
@@ -47,6 +48,32 @@
+
+// Turn A into a string literal without expanding macro definitions
+// (however, if invoked from a macro, macro arguments are expanded).
+#define STRINGIZE_NX(A) #A
```
We have this macro in `Factory.h` and at least one other place in MOOSE but I guess none of that is available here?

permcody commented on framework/src/base/MooseApp.C:
```diff
@@ -47,6 +48,32 @@
+#define STRINGIZE_NX(A) #A
+
+// Turn A into a string literal after macro-expanding it.
+#define STRINGIZE(A) STRINGIZE_NX(A)
```
This one may be new. Maybe we just need to put these in `MooseTypes.h` and quite replicating them around.

jwpeterson replied:
Yes, this is not already in MOOSE and I need both of them in order for this to work.  If you want me to add it somewhere else instead, I can.

permcody replied:
Not strictly needed to merge this. However, if you see an opportunity to clean up the existing macros and we are including some common header in the right places already, why not?

permcody commented:
This looks good so far. I think we really should extend the libMesh checks so that we can see the compiler capabilities with or without the C++11 standard enabled. That will give us a good tool for figuring out compatibility going forward. I think we should shoot to have this done before the training. Perhaps you'll find it useful when you are on the road. @friedmud certainly will when he's doing the East Coast training right after UNM.

jwpeterson commented:
> This looks good so far. I think we really should extend the libMesh checks so that we can see the compiler capabilities with or without the C++11 standard enabled.

OK, that's what I'm working on at the moment.

friedmud commented:
So this is going to come out at the end of every simulation?  That really seems heavy handed.  Why not have a command-line switch to turn this ON?  It's just extra noise that people will need to ignore.

I still really think that the best thing to do is have a script that will submit this information to us.  We can ask the users to run it.

permcody commented:
The plan is to only print this on machines where all the answers are NOT "yes". If our target users are really ready for C++11 (which you use every opportunity to push for :smile:) then nobody will see this.

 Is that really heavy handed? :smirk:

andrsd commented:
We can also add a command line switch that will do this check and prints out the report. Simply redirect the output to a file and let them send the file to us.

dschwen commented:
![](http://cdn.meme.am/instances2/500x/4784648.jpg)

permcody commented:
![image](https://cloud.githubusercontent.com/assets/3036960/13793399/52aa1500-eabc-11e5-95de-aa122a238aff.png)

jwpeterson committed:
Update libmesh.

Short summary of changes in this update:
* Enable FETypes of arbitrary order (larger than the largest Order enumberation).
* Properly free communicators created by calling split().
* Fixes for adaptive p and h-p refinement.
* More thorough testing for C++11 threads (std::atomic, std::mutex, etc).
* Fix issue with VTK autoconf tests that caused test programs to fail linking.
* Enhance and expand C++11 configure tests.
* Make TypeTensor row() const.
* Optimize/fix issues with constant monomial projection.
* Make --enable-unique-ptr is now the default, MOOSE has been using this for a while already.
* Allow threading directly over std::vectors (without copying).
* Fix issue with XDA I/O on highly-refined meshes.

Refs #000.

friedmud commented:
I was thinking it was "heavy handed" because it's going to come out after every run.  Is there no way to provide this report when they _compile_?  Seeing it at the and of `make` might make more sense...

permcody commented:
We talked about that. I'm just concerned about missing this information. I
think there's a high chance that users will ignore messages at the end of
make. Crap half the time users move on even where there are a string of
errors at the end.

I think setting an environment variable or touching a file to disable
runtime messages might be the better path.

Basically it boils down to me not trusting the user to take action unless
it is slightly annoying. If we don't get feedback we run the risk of
finding out after we pull the switch which might mean reversing course in
our development plan. I really don't want that to happen.

If you think this is widespread well then we aren't even close to moving to
C++11.
On Wed, Mar 16, 2016 at 10:03 AM Derek Gaston <notifications@github.com>
wrote:

> I was thinking it was "heavy handed" because it's going to come out after
> every run. Is there no way to provide this report when they *compile*?
> Seeing it at the and of make might make more sense...
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6581#issuecomment-197398852>
>

jwpeterson commented:
The latest version adds a command line argument `--check-cxx11` which controls the printing.  So, instead of having a separate script, we can just have peopel run run e.g. `./moose_test-opt --check-cxx11` to print the C++11 stuff out.  As @andrsd suggested, the stuff also gets printed to a file with instructions to send it to the mailing list.

Some examples of what it prints on representative systems:

OSX/Clang 3.7.0:
```
Compiler C++11 compatibility report: 
  Basic functionality:     yes
  Alias declarations:      yes
  auto keyword:            yes
  constexpr keyword:       yes
  decltype keyword:        yes
  Deleted functions:       yes
  'final' keyword:         yes
  Initializer lists:       yes
  Lambdas:                 yes
  'move' keyword:          yes
  'nullptr' keyword:       yes
  'override' keyword:      yes
  <regex>:                 yes
  rvalue references:       yes
  std::shared_ptr:         yes
  <thread>:                yes
  <type_traits>:           yes
  std::unique_ptr:         yes
  Variadic templates:      yes
```

Linux/GCC 4.6.3:
```
Compiler C++11 compatibility report: 
  Basic functionality:     no
  Alias declarations:      no
  auto keyword:            yes
  constexpr keyword:       yes
  decltype keyword:        yes
  Deleted functions:       yes
  'final' keyword:         no
  Initializer lists:       yes
  Lambdas:                 yes
  'move' keyword:          yes
  'nullptr' keyword:       yes
  'override' keyword:      no
  <regex>:                 no
  rvalue references:       yes
  std::shared_ptr:         yes
  <thread>:                no
  <type_traits>:           no
  std::unique_ptr:         yes
  Variadic templates:      yes
```

By the way @permcody, I realized there was already the `QUOTE` macro which does what I needed to do, so I used that instead.

> Seeing it at the and of make might make more sense...

I could make it print out in the libmesh config summary, but this has the same issue of being buried hundreds of lines deep in your terminal output...

Doing it from `make` would likely mean a separate script again, which would certainly be possible.  But, I thought when @friedmud first came up with this idea, it was specifically to be a runtime feature, so we could start actively "alerting" people to the fact that their compiler did not support certain aspects of C++11.

friedmud commented:
I personally don't think we need any feedback.  We just nee to draw a line in the sand and say what we support.  Just like with OSX... we only support the two most recent versions.  We need to do the same for Linux.  We cannot sit around and wait on everyone to willingly upgrade their stuff.  We have to move forward.

Have you checked to see what comes up as true/false from Ubuntu 14.04?  In my mind we should support the latest two LTS releases from Ubuntu.  Anything older than that needs to upgrade.

Why do you think that spamming people when they run the code is going to make them any more likely to give you feedback?  They're going to ignore these messages just like they ignore the hundreds of other messages they see all the time.  The best way to get feedback is to make it simple to provide feedback.  If people have to copy/paste this into an email (and send it where?   moose-users?  So we're going to spam that now too?) they simply won't do it.

If you want feedback a simple script they can run that gathers up some machine info and auto-submits it to our database is going to be FAR more effective (and less annoying).

jwpeterson commented:
> Have you checked to see what comes up as true/false from Ubuntu 14.04?

I will do that now.  The output above is from Ubuntu 12.04 on rod.

YaqiWang commented:
Seems you guys need to create a separate issue discussing c++11...

jwpeterson commented:
> I will do that now. The output above is from Ubuntu 12.04 on rod.

The compiler on 14.04 is GCC 4.8.4
```
> g++ --version
g++ (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4
```

And it has support for everything except `<type_traints>` and `std::regex` which aren't of huge importance to us.
```
checking whether g++ supports C++11 features with -std=gnu++11... yes
<<< Configuring libMesh with methods "opt" >>>
checking for C++11 std::move support... yes
checking for C++11 decltype support... yes
checking for C++11 rvalue references support... yes
checking for C++11 auto keyword support... yes
checking for C++11 lambda support... yes
checking for C++11 constexpr support... yes
checking for C++11 alias declarations support... yes
checking for C++11 std::shared_ptr support with opt flags... yes
checking for C++11 std::unique_ptr support... yes
checking for C++11 std::regex support... no
checking for C++11 override keyword support... yes
checking for C++11 initializer list support... yes
checking for C++11 variadic template support... yes
checking for C++11 <thread> support... yes
checking for C++11 <type_traits> support... no
checking for C++11 std::asinh support in <cmath>... yes
checking for C++11 std::acosh support in <cmath>... yes
checking for C++11 std::atanh support in <cmath>... yes
checking for C++11 std::asinh(complex) support in <complex>... yes
checking for C++11 std::acosh(complex) support in <complex>... yes
checking for C++11 std::atanh(complex) support in <complex>... yes
checking for C++11 deleted functions support... yes
checking for C++11 'final' keyword support... yes
checking for C++11 nullptr support... yes
```

jwpeterson commented:
> We just nee to draw a line in the sand and say what we support. Just like with OSX... we only support the two most recent versions.

Linux is not like OSX though.  There are clusters at partner institutions which either aren't or can't be upgraded.  I'm not saying we shouldn't cut them off, just that we should be aware how many people we're cutting off...

friedmud commented:
All of the clusters I've seen recently, while possibly running older OS's... are all running newer compilers.... and have even newer compilers installed often.

Looking at 12.04 and 14.04 I would say that we could start using many features of C++11 now.  Lamdas and move operators for instance.  We could start there and see where we go from there.

jwpeterson commented:
Another data point: the build boxes all use GCC 4.9.1 from the `moose-dev-gcc` module:

```
g++ (GCC) 4.9.1
Copyright (C) 2014 Free Software Foundation, Inc.
```

Looks like they got `std::regex` working but not `<type_traits>` yet.
```
checking whether ccache mpicxx supports C++11 features with -std=gnu++11... yes
<<< Configuring libMesh with methods "opt" >>>
checking for C++11 std::move support... yes
checking for C++11 decltype support... yes
checking for C++11 rvalue references support... yes
checking for C++11 auto keyword support... yes
checking for C++11 lambda support... yes
checking for C++11 constexpr support... yes
checking for C++11 alias declarations support... yes
checking for C++11 std::shared_ptr support with opt flags... yes
checking for C++11 std::unique_ptr support... yes
checking for C++11 std::regex support... yes
checking for C++11 override keyword support... yes
checking for C++11 initializer list support... yes
checking for C++11 variadic template support... yes
checking for C++11 <thread> support... yes
checking for C++11 <type_traits> support... no
checking for C++11 std::asinh support in <cmath>... yes
checking for C++11 std::acosh support in <cmath>... yes
checking for C++11 std::atanh support in <cmath>... yes
checking for C++11 std::asinh(complex) support in <complex>... yes
checking for C++11 std::acosh(complex) support in <complex>... yes
checking for C++11 std::atanh(complex) support in <complex>... yes
checking for C++11 deleted functions support... yes
checking for C++11 'final' keyword support... yes
checking for C++11 nullptr support... yes
```

friedmud commented:
2 other things I would vote for right now: `auto` and the "for each" looping syntax.

Do we have a check for that looping syntax?

jwpeterson commented:
> Do we have a check for that looping syntax?

Oh yeah, you are talking about the range-based for loops, like `for(auto i: v)`.  I don't have a test for that, but I will add one.

friedmud commented:
Yes.  They are insanely awesome.  Combined with Lambdas and our threading infrastructure you can do some cool stuff very succinctly...

permcody commented:
I still think you are sending mixed messages here:

> I personally don't think we need any feedback. We just nee to draw a line in the sand and say what we support. Just like with OSX... we only support the two most recent versions. We need to do the same for Linux. We cannot sit around and wait on everyone to willingly upgrade their stuff. We have to move forward.

> All of the clusters I've seen recently, while possibly running older OS's... are all running newer compilers.... and have even newer compilers installed often.

Ok, fine, if you REALLY believe people aren't running old compilers, what's the harm of printing out messages? I kind of think you want to deal with this with the "head in the sand" approach. 

I'd rather make it a little painful, if it pisses a lot of people off, then we'll pull it right back out and decide whether we can realistically move forward or not, (probably not). If however we put this in and we get two responses back from people running old clusters with RedHat at some University that isn't part of our target customer base (or more importantly BISON's customer), well still pull it out and move forward.

jwpeterson commented:
OK, latest version also has checks for range-based for loop.  

As far as the plan for moving forward: @permcody and I were thinking about having multiple phases:

Phase 1: The "opt-in" period.  During this time, we turn on --enable-cxx11 in `update_and_rebuild_libmesh.sh`, and advertise (twitter/blog/email/etc.) that we'd like people to run `./moose_test-opt --check-cxx11` and send in the results to us.  We could have them go to moose-devel if you'd prefer not to spam the users list.  We'll use these initial results as a survey to find out what kind of compilers people are actually using (and in what numbers) so we have better-than-anecdotal evidence about what's out there.

Phase 2: Depending on the results of Phase 1, we'd then turn on the "annoying" version of the check, which would, by default, print out a slightly different message at the end of simulations if your compiler does not support the full C++11 subset (let's say lambdas, auto, range-for, std::move) that we decide to require.

Phase 3: In Phase 3, we actually start putting lambdas, auto, range-for, etc. into MOOSE and officially requiring C++11.

friedmud commented:
I'm ok with the plan.  You guys can decide on the particulars (which email list, etc.).

One thing though: will the `--check-cxx11` option print out _which_ compiler they are using?  That seems like it would be really important information.  I would also really want to get a `uname -a` type of info...

jwpeterson commented:
> One thing though: will the --check-cxx11 option print out which compiler they are using? That seems like it would be really important information. I would also really want to get a uname -a type of info...

Yeah, good point.  We could probably make system calls to `$CXX --version` and `uname -a` for these.

jwpeterson committed:
Print out compiler and system information.

jwpeterson commented:
Printout now looks like this:
```
----------------------------------------
C++11 compiler compatibility report: 
OS: #36-Ubuntu SMP Tue Apr 10 20:39:51 UTC 2012
Compiler: GCC 4.6.3
----------------------------------------
  Basic functionality:     no
  Alias declarations:      no
  auto keyword:            yes
  constexpr keyword:       yes
  decltype keyword:        yes
  Deleted functions:       yes
  'final' keyword:         no
  Initializer lists:       yes
  Lambdas:                 yes
  'move' keyword:          yes
  'nullptr' keyword:       yes
  'override' keyword:      no
  range-based for:         no
  <regex>:                 no
  rvalue references:       yes
  std::shared_ptr:         yes
  <thread>:                no
  <type_traits>:           no
  std::unique_ptr:         yes
  Variadic templates:      yes
```

dschwen commented:
So... when do we get our test boxes to support C++11? I'd like to go ahead with Magpie, but... https://www.moosebuild.org/job/15044/

permcody commented:
When HPC gets our remaining boxes up. I don't think we have the capacity to support C++11 testing until that happens.

jwpeterson commented:
It's up to you guys if you want to wait on libMesh/libmesh#871 to be merged.  If you wait, I'll update this PR with the new libmesh hash when I get back from vacation.

jwpeterson commented:
The libmesh update now contains libMesh/libmesh#871 plus a small fix that @dschwen committed recently, so I don't think we are waiting for anything else.  Is there a "Really Ready to Merge" label?

permcody commented:
Yes, Let's get this in

On Tue, Mar 29, 2016 at 8:29 AM John W. Peterson <notifications@github.com>
wrote:

> The libmesh update now contains libMesh/libmesh#871
> <https://github.com/libMesh/libmesh/pull/871> plus a small fix that
> @dschwen <https://github.com/dschwen> committed recently, so I don't
> think we are waiting for anything else. Is there a "Really Ready to Merge"
> label?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6581#issuecomment-202921411>
>

permcody commented:
:+1:
